File: tests/bench/crosslang/c/http.c
/* Порт tests/bench/programs/HttpBench.eat 1:1 (методика CROSSLANG):
 * тот же HTTP-парсер, что lib/Http.eat — request-line + заголовки в
 * фиксированные буферы, срезы-Header, OWS-трим, 400/413/431, мини-
 * роутер и keep-alive-решение; та же чек-сумма. */
#include <stdint.h>
#include <stdio.h>
#include <string.h>

static const uint32_t REPEAT = 1;

#define HTTP_NONE 0xffffffffu
#define PARSE_MORE 0u
#define PARSE_DONE 1u

typedef struct {
    uint32_t ns, nl, vs, vl;
} Header;

typedef struct {
    uint8_t raw[8192];
    uint32_t n, state, ls, ms, ml, ps, pl, vs, vl, nh, err;
    Header hdr[64];
} Req;

static void req_new(Req *r) {
    memset(r, 0, sizeof(*r));
}

/* Зеркало Req.reset (OPTIMIZATIONS §8.2): скалярные поля в 0,
 * raw/hdr не трогаются — кадр переиспользуется между запросами. */
static void req_reset(Req *r) {
    r->n = 0;
    r->state = 0;
    r->ls = 0;
    r->ms = 0;
    r->ml = 0;
    r->ps = 0;
    r->pl = 0;
    r->vs = 0;
    r->vl = 0;
    r->nh = 0;
    r->err = 0;
}

static uint32_t find_sp(const Req *r, uint32_t b0, uint32_t e0) {
    for (uint32_t i = b0; i < e0; i++) {
        if (r->raw[i] == 32) {
            return i;
        }
    }
    return HTTP_NONE;
}

static uint32_t parse_reqline(Req *r, uint32_t ls2, uint32_t le) {
    uint32_t sp1 = find_sp(r, ls2, le);
    uint32_t sp2 = HTTP_NONE;
    int bad = sp1 == HTTP_NONE;
    if (!bad) {
        sp2 = find_sp(r, sp1 + 1, le);
        bad = sp2 == HTTP_NONE;
    }
    if (!bad) {
        bad = sp1 == ls2 || sp2 == sp1 + 1 || sp2 + 1 == le ||
              find_sp(r, sp2 + 1, le) != HTTP_NONE;
    }
    if (bad) {
        r->state = 3;
        r->err = 400;
        return 400;
    }
    r->ms = ls2;
    r->ml = sp1 - ls2;
    r->ps = sp1 + 1;
    r->pl = sp2 - sp1 - 1;
    r->vs = sp2 + 1;
    r->vl = le - sp2 - 1;
    r->state = 1;
    return PARSE_MORE;
}

static uint32_t parse_header(Req *r, uint32_t ls2, uint32_t le) {
    if (r->nh >= 64) {
        r->state = 3;
        r->err = 431;
        return 431;
    }
    uint32_t colon = HTTP_NONE;
    for (uint32_t i = ls2; i < le; i++) {
        if (r->raw[i] == 58) {
            colon = i;
            break;
        }
    }
    if (colon == HTTP_NONE || colon == ls2) {
        r->state = 3;
        r->err = 400;
        return 400;
    }
    uint32_t vs2 = colon + 1, vl2 = le - vs2;
    while (vl2 > 0 && (r->raw[vs2] == 32 || r->raw[vs2] == 9)) {
        vs2++;
        vl2--;
    }
    while (vl2 > 0) {
        uint8_t tb = r->raw[vs2 + vl2 - 1];
        if (tb != 32 && tb != 9) {
            break;
        }
        vl2--;
    }
    r->hdr[r->nh].ns = ls2;
    r->hdr[r->nh].nl = colon - ls2;
    r->hdr[r->nh].vs = vs2;
    r->hdr[r->nh].vl = vl2;
    r->nh++;
    return PARSE_MORE;
}

static uint32_t push_byte(Req *r, uint8_t b) {
    if (r->state == 2) {
        return PARSE_DONE;
    }
    if (r->state == 3) {
        return r->err;
    }
    if (r->n >= 8192) {
        r->state = 3;
        r->err = 413;
        return 413;
    }
    r->raw[r->n++] = b;
    if (b != 10) {
        return PARSE_MORE;
    }
    uint32_t le = r->n - 1;
    if (le > r->ls && le >= 1 && r->raw[le - 1] == 13) {
        le--;
    }
    uint32_t ls2 = r->ls;
    r->ls = r->n;
    if (r->state == 0) {
        return parse_reqline(r, ls2, le);
    }
    if (le == ls2) {
        r->state = 2;
        return PARSE_DONE;
    }
    return parse_header(r, ls2, le);
}

static uint32_t feed_line(Req *r, const char *s) {
    uint32_t st = PARSE_MORE;
    for (const char *p = s; *p; p++) {
        st = push_byte(r, (uint8_t)*p);
    }
    st = push_byte(r, 13);
    st = push_byte(r, 10);
    return st;
}

static uint8_t lower(uint8_t b) {
    return (b >= 65 && b <= 90) ? (uint8_t)(b + 32) : b;
}

static int span_is(const Req *r, uint32_t s, uint32_t l, const char *kw) {
    if (l != (uint32_t)strlen(kw)) {
        return 0;
    }
    for (uint32_t i = 0; i < l; i++) {
        if (r->raw[s + i] != (uint8_t)kw[i]) {
            return 0;
        }
    }
    return 1;
}

static int span_is_ci(const Req *r, uint32_t s, uint32_t l, const char *kw) {
    if (l != (uint32_t)strlen(kw)) {
        return 0;
    }
    for (uint32_t i = 0; i < l; i++) {
        if (lower(r->raw[s + i]) != lower((uint8_t)kw[i])) {
            return 0;
        }
    }
    return 1;
}

static int method_is(const Req *r, const char *m) {
    return span_is(r, r->ms, r->ml, m);
}

static int path_is(const Req *r, const char *p) {
    return span_is(r, r->ps, r->pl, p);
}

static int version_is(const Req *r, const char *v) {
    return span_is(r, r->vs, r->vl, v);
}

static int path_starts(const Req *r, const char *pre) {
    uint32_t pl = (uint32_t)strlen(pre);
    if (r->pl < pl) {
        return 0;
    }
    return span_is(r, r->ps, pl, pre);
}

/* хвост пути после skip байт; -1 — нет или длиннее 64 (None) */
static int32_t path_param_len(const Req *r, uint32_t skip) {
    if (r->pl <= skip || r->pl - skip > 64) {
        return -1;
    }
    return (int32_t)(r->pl - skip);
}

static uint32_t find_header(const Req *r, const char *name) {
    for (uint32_t i = 0; i < r->nh; i++) {
        if (span_is_ci(r, r->hdr[i].ns, r->hdr[i].nl, name)) {
            return i;
        }
    }
    return HTTP_NONE;
}

static int header_val_is(const Req *r, uint32_t i, const char *v) {
    return span_is_ci(r, r->hdr[i].vs, r->hdr[i].vl, v);
}

static int wants_close(const Req *r) {
    uint32_t c = find_header(r, "connection");
    if (c != HTTP_NONE && c < 64) {
        if (header_val_is(r, c, "close")) {
            return 1;
        }
        if (header_val_is(r, c, "keep-alive")) {
            return 0;
        }
    }
    return version_is(r, "HTTP/1.0");
}

static uint32_t route_code(const Req *r) {
    if (path_is(r, "/")) {
        return 200;
    }
    if (path_starts(r, "/greet/")) {
        return 201;
    }
    return 404;
}

static uint32_t profile_a(Req *r, uint32_t k) {
    char line[64];
    req_reset(r);
    snprintf(line, sizeof(line), "GET /greet/user%u HTTP/1.1", k);
    feed_line(r, line);
    feed_line(r, "Host: bench.local");
    feed_line(r, "User-Agent: eat-bench/1.0");
    feed_line(r, "Accept: */*");
    feed_line(r, "Connection: keep-alive");
    uint32_t st = feed_line(r, "");
    uint32_t acc = st * 7 + r->nh;
    if (method_is(r, "GET")) {
        acc += route_code(r);
    }
    int32_t pl = path_param_len(r, 7);
    if (pl >= 0) {
        acc += (uint32_t)pl;
    }
    if (!wants_close(r)) {
        acc += 1;
    }
    uint32_t c = find_header(r, "connection");
    if (c != HTTP_NONE) {
        acc += c;
    }
    return acc;
}

static uint32_t profile_b(Req *r, uint32_t k) {
    char line[64];
    req_reset(r);
    feed_line(r, "POST /api/items HTTP/1.1");
    feed_line(r, "Content-Type: application/json");
    snprintf(line, sizeof(line), "X-Trace-Id: t-%u", k);
    feed_line(r, line);
    feed_line(r, "Connection: close");
    uint32_t st = feed_line(r, "");
    uint32_t acc = st * 7 + r->nh + route_code(r);
    if (wants_close(r)) {
        acc += 2;
    }
    uint32_t t = find_header(r, "x-trace-id");
    if (t != HTTP_NONE && t < 64) {
        acc += r->hdr[t].vl;
    }
    return acc;
}

static uint32_t profile_c(Req *r, uint32_t k) {
    char line[64];
    req_reset(r);
    snprintf(line, sizeof(line), "BROKEN-%u", k);
    return feed_line(r, line);
}

static uint32_t profile_d(Req *r) {
    req_reset(r);
    feed_line(r, "GET /nope HTTP/1.0");
    uint32_t st = feed_line(r, "");
    uint32_t acc = st * 7 + route_code(r);
    if (wants_close(r)) {
        acc += 3;
    }
    return acc;
}

int main(void) {
    uint32_t acc = 0;
    Req r;
    req_new(&r);
    for (uint32_t rep = 0; rep < REPEAT; rep++) {
        for (uint32_t k = 0; k < 500; k++) {
            acc = (acc * 31 + profile_a(&r, k)) % 65536;
            acc = (acc * 31 + profile_b(&r, k)) % 65536;
            acc = (acc * 31 + profile_c(&r, k)) % 65536;
            acc = (acc * 31 + profile_d(&r)) % 65536;
        }
    }
    printf("checksum %u\n", acc);
    return 0;
}

File: tests/bench/crosslang/rust/http.rs
// Порт tests/bench/programs/HttpBench.eat 1:1 (методика CROSSLANG):
// тот же HTTP-парсер, что lib/Http.eat, та же чек-сумма.

const REPEAT: u32 = 1;

const HTTP_NONE: u32 = 0xffff_ffff;
const PARSE_MORE: u32 = 0;
const PARSE_DONE: u32 = 1;

#[derive(Clone, Copy, Default)]
struct Header {
    ns: u32,
    nl: u32,
    vs: u32,
    vl: u32,
}

struct Req {
    raw: [u8; 8192],
    n: u32,
    state: u32,
    ls: u32,
    ms: u32,
    ml: u32,
    ps: u32,
    pl: u32,
    vs: u32,
    vl: u32,
    hdr: [Header; 64],
    nh: u32,
    err: u32,
}

impl Req {
    fn new() -> Req {
        Req {
            raw: [0; 8192],
            n: 0,
            state: 0,
            ls: 0,
            ms: 0,
            ml: 0,
            ps: 0,
            pl: 0,
            vs: 0,
            vl: 0,
            hdr: [Header::default(); 64],
            nh: 0,
            err: 0,
        }
    }

    fn find_sp(&self, b0: u32, e0: u32) -> u32 {
        for i in b0..e0 {
            if self.raw[i as usize] == 32 {
                return i;
            }
        }
        HTTP_NONE
    }

    fn parse_reqline(&mut self, ls2: u32, le: u32) -> u32 {
        let sp1 = self.find_sp(ls2, le);
        let mut sp2 = HTTP_NONE;
        let mut bad = sp1 == HTTP_NONE;
        if !bad {
            sp2 = self.find_sp(sp1 + 1, le);
            bad = sp2 == HTTP_NONE;
        }
        if !bad {
            bad = sp1 == ls2
                || sp2 == sp1 + 1
                || sp2 + 1 == le
                || self.find_sp(sp2 + 1, le) != HTTP_NONE;
        }
        if bad {
            self.state = 3;
            self.err = 400;
            return 400;
        }
        self.ms = ls2;
        self.ml = sp1 - ls2;
        self.ps = sp1 + 1;
        self.pl = sp2 - sp1 - 1;
        self.vs = sp2 + 1;
        self.vl = le - sp2 - 1;
        self.state = 1;
        PARSE_MORE
    }

    fn parse_header(&mut self, ls2: u32, le: u32) -> u32 {
        if self.nh >= 64 {
            self.state = 3;
            self.err = 431;
            return 431;
        }
        let mut colon = HTTP_NONE;
        for i in ls2..le {
            if self.raw[i as usize] == 58 {
                colon = i;
                break;
            }
        }
        if colon == HTTP_NONE || colon == ls2 {
            self.state = 3;
            self.err = 400;
            return 400;
        }
        let mut vs2 = colon + 1;
        let mut vl2 = le - vs2;
        while vl2 > 0 {
            let b = self.raw[vs2 as usize];
            if b != 32 && b != 9 {
                break;
            }
            vs2 += 1;
            vl2 -= 1;
        }
        while vl2 > 0 {
            let tb = self.raw[(vs2 + vl2 - 1) as usize];
            if tb != 32 && tb != 9 {
                break;
            }
            vl2 -= 1;
        }
        self.hdr[self.nh as usize] = Header {
            ns: ls2,
            nl: colon - ls2,
            vs: vs2,
            vl: vl2,
        };
        self.nh += 1;
        PARSE_MORE
    }

    fn push_byte(&mut self, b: u8) -> u32 {
        if self.state == 2 {
            return PARSE_DONE;
        }
        if self.state == 3 {
            return self.err;
        }
        if self.n >= 8192 {
            self.state = 3;
            self.err = 413;
            return 413;
        }
        self.raw[self.n as usize] = b;
        self.n += 1;
        if b != 10 {
            return PARSE_MORE;
        }
        let mut le = self.n - 1;
        if le > self.ls && le >= 1 && self.raw[(le - 1) as usize] == 13 {
            le -= 1;
        }
        let ls2 = self.ls;
        self.ls = self.n;
        if self.state == 0 {
            return self.parse_reqline(ls2, le);
        }
        if le == ls2 {
            self.state = 2;
            return PARSE_DONE;
        }
        self.parse_header(ls2, le)
    }

    fn feed_line(&mut self, s: &str) -> u32 {
        for &b in s.as_bytes() {
            self.push_byte(b);
        }
        self.push_byte(13);
        self.push_byte(10)
    }

    fn span_is(&self, s: u32, l: u32, kw: &str) -> bool {
        let kb = kw.as_bytes();
        if l as usize != kb.len() {
            return false;
        }
        for i in 0..l {
            if self.raw[(s + i) as usize] != kb[i as usize] {
                return false;
            }
        }
        true
    }

    fn span_is_ci(&self, s: u32, l: u32, kw: &str) -> bool {
        let kb = kw.as_bytes();
        if l as usize != kb.len() {
            return false;
        }
        for i in 0..l {
            if lower(self.raw[(s + i) as usize]) != lower(kb[i as usize]) {
                return false;
            }
        }
        true
    }

    fn method_is(&self, m: &str) -> bool {
        self.span_is(self.ms, self.ml, m)
    }

    fn path_is(&self, p: &str) -> bool {
        self.span_is(self.ps, self.pl, p)
    }

    fn version_is(&self, v: &str) -> bool {
        self.span_is(self.vs, self.vl, v)
    }

    fn path_starts(&self, pre: &str) -> bool {
        if (self.pl as usize) < pre.len() {
            return false;
        }
        self.span_is(self.ps, pre.len() as u32, pre)
    }

    // хвост пути после skip байт; -1 — нет или длиннее 64 (None)
    fn path_param_len(&self, skip: u32) -> i32 {
        if self.pl <= skip || self.pl - skip > 64 {
            return -1;
        }
        (self.pl - skip) as i32
    }

    fn find_header(&self, name: &str) -> u32 {
        for i in 0..self.nh {
            let h = self.hdr[i as usize];
            if self.span_is_ci(h.ns, h.nl, name) {
                return i;
            }
        }
        HTTP_NONE
    }

    fn header_val_is(&self, i: u32, v: &str) -> bool {
        let h = self.hdr[i as usize];
        self.span_is_ci(h.vs, h.vl, v)
    }

    fn wants_close(&self) -> bool {
        let c = self.find_header("connection");
        if c != HTTP_NONE && c < 64 {
            if self.header_val_is(c, "close") {
                return true;
            }
            if self.header_val_is(c, "keep-alive") {
                return false;
            }
        }
        self.version_is("HTTP/1.0")
    }
}

fn lower(b: u8) -> u8 {
    if (65..=90).contains(&b) {
        b + 32
    } else {
        b
    }
}

fn route_code(r: &Req) -> u32 {
    if r.path_is("/") {
        return 200;
    }
    if r.path_starts("/greet/") {
        return 201;
    }
    404
}

fn profile_a(k: u32) -> u32 {
    let mut r = Req::new();
    r.feed_line(&format!("GET /greet/user{} HTTP/1.1", k));
    r.feed_line("Host: bench.local");
    r.feed_line("User-Agent: eat-bench/1.0");
    r.feed_line("Accept: */*");
    r.feed_line("Connection: keep-alive");
    let st = r.feed_line("");
    let mut acc = st * 7 + r.nh;
    if r.method_is("GET") {
        acc += route_code(&r);
    }
    let pl = r.path_param_len(7);
    if pl >= 0 {
        acc += pl as u32;
    }
    if !r.wants_close() {
        acc += 1;
    }
    let c = r.find_header("connection");
    if c != HTTP_NONE {
        acc += c;
    }
    acc
}

fn profile_b(k: u32) -> u32 {
    let mut r = Req::new();
    r.feed_line("POST /api/items HTTP/1.1");
    r.feed_line("Content-Type: application/json");
    r.feed_line(&format!("X-Trace-Id: t-{}", k));
    r.feed_line("Connection: close");
    let st = r.feed_line("");
    let mut acc = st * 7 + r.nh + route_code(&r);
    if r.wants_close() {
        acc += 2;
    }
    let t = r.find_header("x-trace-id");
    if t != HTTP_NONE && t < 64 {
        acc += r.hdr[t as usize].vl;
    }
    acc
}

fn profile_c(k: u32) -> u32 {
    let mut r = Req::new();
    r.feed_line(&format!("BROKEN-{}", k))
}

fn profile_d() -> u32 {
    let mut r = Req::new();
    r.feed_line("GET /nope HTTP/1.0");
    let st = r.feed_line("");
    let mut acc = st * 7 + route_code(&r);
    if r.wants_close() {
        acc += 3;
    }
    acc
}

fn main() {
    let mut acc: u32 = 0;
    for _ in 0..REPEAT {
        for k in 0..500u32 {
            acc = (acc * 31 + profile_a(k)) % 65536;
            acc = (acc * 31 + profile_b(k)) % 65536;
            acc = (acc * 31 + profile_c(k)) % 65536;
            acc = (acc * 31 + profile_d()) % 65536;
        }
    }
    println!("checksum {}", acc);
}
